### PR TITLE
Elrond to Ethereum state machine

### PR DESCRIPTION
**Pull request by dragos-rebegea**
-add Elrond to Ethereum state machine

### CODE REVIEW

File: ethToElrond/v2/elrondToEth/steps/step07SignProposedSetStatus.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+func (step *signProposedSetStatusStep) Execute(ctx context.Context) (core.StepIdentifier, error) {
+	batch := step.bridge.GetStoredBatch()
+	if batch == nil {
+		step.bridge.GetLogger().Debug("no batch found")
+		return elrondToEth.GettingPendingBatchFromElrond, nil
+	}
+
+	actionID, err := step.bridge.GetAndStoreActionIDForSetStatusFromElrond(ctx)
```
Comment (iulianpascalau) [inline ethToElrond/v2/elrondToEth/steps/step07SignProposedSetStatus.go:22]: there will be an intermediate step here, forgot about it :(
The step in which we get the statuses of the transfers executed on Ethereum and apply them on the current stored batch.

File: ethToElrond/v2/elrondToEth/steps/step08WaitForQuorumOnSetStatus.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+	bridge elrondToEth.ElrondToEthBridge
+}
+
+func (step *waitForQuorumOnSetStatus) Execute(ctx context.Context) (core.StepIdentifier, error) {
+	if step.bridge.ProcessMaxRetriesOnElrond() {
+		step.bridge.GetLogger().Debug("max number of retries reached, resetting counter")
+		return elrondToEth.GettingPendingBatchFromElrond, nil
+	}
+
+	isQuorumReached, err := step.bridge.IsQuorumReachedOnElrond(ctx)
```
Comment (iulianpascalau) [inline ethToElrond/v2/elrondToEth/steps/step08WaitForQuorumOnSetStatus.go:20]: ProcessQuorumReachedOnElrond *

File: ethToElrond/v2/errors.go [hunk 3]
```diff
@@ -2,11 +2,16 @@ package v2
...
 
-var (
-	errNilBatch            = errors.New("nil batch")
-	errInvalidDepositNonce = errors.New("invalid deposit nonce")
-	errNilLogger           = errors.New("nil logger")
-	errNilElrondClient     = errors.New("nil Elrond client")
-	errNilEthereumClient   = errors.New("nil Ethereum client")
-	errNilTopologyProvider = errors.New("nil topology provider")
-)
+// TODO add comments
```
Comment (iulianpascalau) [inline ethToElrond/v2/errors.go:5]: Implement TODOs :D
Also, why do we need them exported?

File: ethToElrond/v2/ethToElrond/steps/step03SignProposedTransfer.go [hunk 4]
```diff
@@ -26,7 +25,7 @@ func (step *signProposedTransferStep) Execute(ctx context.Context) (core.StepIde
 		step.bridge.GetLogger().Error("error fetching action ID", "batch ID", batch.ID, "error", err)
 		return ethToElrond.GettingPendingBatchFromEthereum, nil
 	}
-	if actionID == invalidActionID {
+	if actionID == v2.InvalidActionID {
```
Comment (iulianpascalau) [inline ethToElrond/v2/ethToElrond/steps/step03SignProposedTransfer.go:28]: :+1:

**Review by iulianpascalau — CHANGES_REQUESTED** [review]
Good impl :+1:

File: ethToElrond/v2/errors.go (hunk 3 above)
Comment (dragos-rebegea) [reply ethToElrond/v2/errors.go:5]: those are defined in /v2/errors and we use them in /v2/elrondToEth/steps

File: ethToElrond/v2/elrondToEth/interface.go [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+import (
+	"context"
+
+	"github.com/ElrondNetwork/elrond-eth-bridge/clients"
+	logger "github.com/ElrondNetwork/elrond-go-logger"
+)
+
+// ElrondToEthBridge defines the operations for the Elrond to Ethereum half-bridge
+type ElrondToEthBridge interface {
```
[the remark this answers is not in the document]
Comment (dragos-rebegea) [reply ethToElrond/v2/elrondToEth/interface.go:11]: @sstanculeanu will merge them in a future PR

File: ethToElrond/v2/elrondToEth/constants.go [hunk 6]
```diff
@@ -0,0 +1,50 @@
+package elrondToEth
+
+const (
+	// GettingPendingBatchFromElrond is the step identifier for fetching the pending batch from the Elrond chain
+	GettingPendingBatchFromElrond = "get pending batch from ethereum"
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply ethToElrond/v2/elrondToEth/constants.go:5]: not fixed

File: ethToElrond/v2/elrondToEth/constants.go [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+	SigningProposedTransferOnEthereum = "sign proposed transfer"
+
+	// WaitingForQuorumOnTransfer is the step identifier for waiting until the quorum is reached
+	WaitingForQuorumOnTransfer = "wait for quorum on transfer"
+
+	// PerformingTransfer is the step identifier for performing the transfer on Ethereum
+	PerformingTransfer = "perform transfer"
+
+	// WaitingTransferConfirmation is the step identifier for waiting the transfer confirmation on Ethereum
+	WaitingTransferConfirmation = "wait transfer confirmationg"
```
[the remark this answers is not in the document]
Comment (iulianpascalau) [reply ethToElrond/v2/elrondToEth/constants.go]: not fixed

File: ethToElrond/v2/elrondToEth/steps/step06ResolveSetStatus.go [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+		step.bridge.GetLogger().Debug("nil batch fetched")
+		return elrondToEth.GettingPendingBatchFromElrond, nil
+	}
+
+	statuses, err := step.bridge.GetBatchStatusesFromEthereum(ctx)
+	if err != nil {
+		step.bridge.GetLogger().Error("error while fetching transaction statuses", "error", err)
+		return elrondToEth.GettingPendingBatchFromElrond, nil
+	}
+	if statuses == nil {
```
Comment (iulianpascalau) [inline ethToElrond/v2/elrondToEth/steps/step06ResolveSetStatus.go:36]: We can't have nil here. L36-L39 should be deleted

File: ethToElrond/v2/elrondToEth/steps/step06ResolveSetStatus.go [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+		step.bridge.GetLogger().Error("nil transaction statuses")
+		return elrondToEth.GettingPendingBatchFromElrond, nil
+	}
+
+	for i, transactionStatus := range statuses {
```
Comment (iulianpascalau) [inline ethToElrond/v2/elrondToEth/steps/step06ResolveSetStatus.go:41]: Better for L41-L43: 
```GO
batch.Statuses = statuses
```

File: ethToElrond/v2/elrondToEth/steps/step06ResolveSetStatus.go [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+		batch.Statuses[i] = transactionStatus
+	}
+
+	numStoredBatchDeposits := len(storedBatch.Statuses)
```
Comment (iulianpascalau) [inline ethToElrond/v2/elrondToEth/steps/step06ResolveSetStatus.go:45]: please use directly for L45-L53: 
```GO
step.bridge.ResolveNewDepositsStatuses(len(batch.Statuses))
```

Also, you have a typo on `ResolveNewD**e**positsStatuses`

**Review by iulianpascalau — COMMENTED** [review]
Please do fix the new comments and the unresolved old comments.

File: ethToElrond/v2/elrondToEth/steps/step01GetPending.go [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+
+type getPendingStep struct {
+	bridge elrondToEth.ElrondToEthBridge
+}
+
+func (step *getPendingStep) Execute(ctx context.Context) (core.StepIdentifier, error) {
+	step.bridge.ResetRetriesCountOnEthereum()
+	step.bridge.ResetRetriesCountOnElrond()
+
+	batch, err := step.bridge.GetBatchFromElrond(ctx)
```
Comment (sstanculeanu) [inline ethToElrond/v2/elrondToEth/steps/step01GetPending.go:18]: why not using only one action GetAndStoreBatchFromElrond and then only check err for nil and batch to nil?

File: ethToElrond/v2/elrondToEth/steps/step01GetPending.go [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+	step.bridge.GetLogger().Info("fetched new batch from Elrond " + batch.String())
+
+	wasPerformed, err := step.bridge.WasTransferPerformedOnEthereum(ctx)
+	if err != nil {
+		step.bridge.GetLogger().Error("error determining if transfer was performed or not", "error", err)
+		return elrondToEth.GettingPendingBatchFromElrond, nil
+	}
+	if wasPerformed {
+		step.bridge.GetLogger().Info("transfer performed")
+		return elrondToEth.ProposingSetStatusOnElrond, nil
```
Comment (sstanculeanu) [inline ethToElrond/v2/elrondToEth/steps/step01GetPending.go:49]: shouldn't this go to `elrondToEth.ResolvingSetStatusOnElrond`, same as in `step04PerformTransfer`?

File: ethToElrond/v2/elrondToEth/steps/step04PerformTransfer.go [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+		step.bridge.GetLogger().Error("error determining if transfer was performed or not", "error", err)
+		return elrondToEth.GettingPendingBatchFromElrond, nil
+	}
+
+	if wasPerformed {
+		step.bridge.GetLogger().Info("transfer performed")
+		return elrondToEth.ResolvingSetStatusOnElrond, nil
+	}
+
+	if !step.bridge.MyTurnAsLeader() {
```
Comment (sstanculeanu) [inline ethToElrond/v2/elrondToEth/steps/step04PerformTransfer.go:26]: `if step.bridge.MyTurnAsLeader() {`

File: ethToElrond/v2/elrondToEth/steps/step01GetPending.go (hunk 11 above)
Comment (sstanculeanu) [reply ethToElrond/v2/elrondToEth/steps/step01GetPending.go:18]: nevermind, found the answer in the following steps

**Review by iulianpascalau — APPROVED** [review]

**Review by sstanculeanu — APPROVED** [review]

**Pull request merged**